### DISCUSSION TITLE
SimpleStorage: difference between name and _name

### DISCUSSION THREAD

**Discussion by pharmacistcodes** [smartcontractkit/full-blockchain-solidity-course-py#1078] [Q&A] [answered]
Hello! I'm just wondering why was the variable for name changed... under the struct it was "string name" but when creating the feature to add into the array people it became "string **_name**"

What does the underscore indicate? Is this a completely different variable?

I hope this question makes sense.

**Comment by n4n0b1t3** [comment]
You didn't provide the repository you are referring to. Which lesson are you in right now? Many use a variable with a prefixed underscore for function parameters. Here is an example:
```
make_noise("Wuff");

function make_noise(_noise){
    var animal = "Dog"
    print(animal + " makes " + _noise + " !");
}
```
Here you can see to types of variables:

1. function parameter with underscore
2. function internal variable

Does that answer your question?

**Comment by cromewar** [comment, marked as answer]
Hello @pharmacyhacks 

```Solidity
   struct People {
        uint256 favoriteNumber;
        string name;
    }

 function addPerson(string memory _name, uint256 _favoriteNumber) public {
        people.push(People(_favoriteNumber, _name));
        nameToFavoriteNumber[_name] = _favoriteNumber;
    }
```

I'm pretty much sure you are referring to this code, and the answer is the following.

Using under score on functions or variables means nothing to he compiler itself but is a convention for good practices at coding, the underscore means  that function/variable can only be called from itself, in other words `_name` exists just on the context of the function and can be called just on that specific function.

I does no mean that `name` and `_name` are the same, they are completely different and have different spaces in memory, they actually could be called completely different like
`name` and `_inputVariable`
We use the same name on both just as reference for other people (the ones reading our code)  in order to make it more understandable. 

Also if you use some linting tool like `solhint` it would complain if you don't use the underscore despite the compiler does not care about it, because the correct way to setup a variable like that is with the keyword "internal"

I hope this info might be helpful.

Cheers.

**Comment by pharmacistcodes** [comment]
Ah sorry about that, I'm in Lesson 1 - Adding an Array

@n4n0b1t3 @cromewar thank you for the responses! Yes I was referring to this code:

   struct People {
        uint256 favoriteNumber;
        string name;
    }

 function addPerson(string memory _name, uint256 _favoriteNumber) public {
        people.push(People(_favoriteNumber, _name));
        nameToFavoriteNumber[_name] = _favoriteNumber;
    }

OK underscore means nothing to the compiler and is meant to help understand the code better. 

So the compiler knows that the variables in the struct (favoriteNumber and name) are inputted from the variables from people.push (_favoriteNumber, _name)?

lol maybe I'm overthinking this too much

**Reply by cromewar** [reply]
> So the compiler knows that the variables in the struct (favoriteNumber and name) are inputted from the variables from people.push (_favoriteNumber, _name)?

Yes sir!

**Reply by pharmacistcodes** [reply]
thank you!
